### ISSUE TITLE
Auto-resolve named exports from ES2015 modules

### ISSUE THREAD

**Issue by lencioni** [Galooshi/import-js#199]
It is a common pattern to have a module that exports multiple things, and possibly no default thing. Example:

``` js
export function foo() {
  ...
}

export function bar() {
  ...
}
```

Now, if I want to use one of these functions, I am left in the dust by import-js. It would be nice if we could resolve these named exports automatically.

**Comment by lencioni** [comment]
Short of parsing the entire project into an AST or something, I think that we might be able to solve for the most common cases using some regexes. And, I think that a partial solution that we can iterate on is better than no solution here.

**Comment by trotzig** [comment]
This is related to #184 and #183. I've been thinking about this and I think we could add a cache of some sort to the file system. `.import-js.cache.json` or something. Depending on how long it takes for this cache file to generate, we can either add an explicit command (e.g. `import-js init`) or make it a one-time startup thing. And if we're smart about what we store in this cache, we can probably use a little bit of time on some import calls to check if we need to update the cache. 

I have a hard time seeing a solution that doesn't use a cache. 

When you say "it is common", how common do you mean? I haven't come across many projects that do that.

**Comment by lencioni** [comment]
In the project that I am working in, `ag "export function"` shows up 149 times. I would expect this to become more common as ES2015 modules becomes more widely used.

**Comment by jscheid** [comment]
Hi, this would be great to have!

> Short of parsing the entire project into an AST

Just a heads up that this should be fairly easy with Babel, see https://github.com/babel/babel/tree/master/packages/babel-core#usage

Obviously you'd want to use something like https://github.com/mikeal/watch to keep it snappy.

I'm pretty busy at the moment but I'll try and see if I can whip up a proof of concept, or let me know if there's a better way I can help move this forward. Cheers!

**Comment by lencioni** [comment]
Thanks! We are more concerned with performance than how difficult it is to do, which is why I suggested the regex approach. If you can come up with a method that can do this for a large project without introducing friction for getting import-js set up and without slowing it down too much, that would be really great. 

We are also in the middle of a rewrite from Ruby to JavaScript. If you plan on hacking, you should probably start from the convert-to-js branch.

**Comment by trotzig** [comment]
Yeah, it would be awesome to have more people involved! As @lencioni said, we're currently rewriting the whole project in JavaScript. I wouldn't say we're in the middle, we're almost ready to merge the convert-to-js branch.

**Comment by jscheid** [comment]
Thanks for you quick replies, good to know about the convert-to-js branch. Again, I'm a bit short on time so can't really promise anything at this point.

Regarding performance, I'm using eslint (configured to use `babel-eslint`) via flymake and the feedback it gives me feels pretty much instantaneous. That's parsing the AST after every change I make and then also runs all sorts of analysis on the AST, so if that's fast it should be possible to make this fast when all it has to do is look for export nodes?

**Comment by jscheid** [comment]
I'm not entirely sure but I think flymake also starts a new eslint process all the time, so in a client-server setup it should be even faster.

**Comment by lencioni** [comment]
The main difference is eslint will parse the file you are currently editing to an AST. To implement this feature, you would need to parse every file in a project, including dependencies, which could easily be a hundred thousand files.

**Comment by jscheid** [comment]
Ah, I see now. I thought you were worried about interactive performance, my bad.

How do you define "every file in a project"? I believe the following files should be excluded:
- test cases and example code
- files in transitive dependencies, importing something `from "foo/node_modules/bar"` should not be encouraged
- (arguably, ideally, depending on configuration) minified "dist" files, and for ES6 dependencies either the ES6 source or the "transpiled" code (think `src` vs `lib`)

I happen to be working on a reasonably large JS application that currently has 170 package dependencies, including larger ones such as babel, webpack, d3, lodash. I'm counting around 5k JS files that would need indexing (out of a total 100k JS files in `node_modules`). Are you sure that having to index 100k files is something that would be common? Extrapolating from the above, that size project would have to be listing several thousand direct dependencies in `package.json`.

I've run some quick and dirty benchmarks and it appears that these 5k files can be parsed via babel in about 2-3 minutes wall time on a 2013 MacBook Pro, compared to about 30s via simple regexps.

**Comment by joefiorini** [comment]
Could ternjs could be used to process named exports and file paths? It [knows about ES6 modules](http://ternjs.net/doc/manual.html#plugin_es_modules), and it appears to [index named exports](https://github.com/ternjs/tern/blob/master/plugin/es_modules.js#L54-L73). I don't enough of the specifics behind how tern works under the hood, but seems like it might be useful?

Some might not want to run a tern server at all times, so maybe keep the `namedExports` configuration and (assuming this is a workable options) allow this as an optional upgrade until a better solution presents itself?

**Comment by trotzig** [comment]
That definitely looks promising, thanks for linking to it. It's probably worth looking in to, even if it ends up not being integrated into ImportJS. If we end up using ternjs directly, we might be able to remove the eslint dependency (which we use to find undefined & unused variables).

**Comment by rhettlivingston** [comment]
I am adding this to the Meteor milestone. I believe it a necessity for Meteor projects. 

Meteor projects that follow the Meteor Guide use named exports almost exclusively and multiple exports-per-module is the norm, especially from API modules. Thus, the current logic assuming that a module will have a filename related to the variable being imported is ineffective more often than not. Without automatic detection of named exports, the importjs configuration file would require maintenance throughout development.

That said, I am not eager to jump straight to the full blown case of implementing this via an AST and utilizing a cache or utilizing ternjs (which uses an AST and cache... and much more). We will likely end up there, but I'd like to get there in stages. I feel that a first cut that supports only ES2015 modules and utilizes regexes for speed would be effective for Meteor's use case.

**Comment by rhettlivingston** [comment]
Stumbled on [this node enhancement proposal for directly supporting ES6 modules](https://github.com/nodejs/node-eps/blob/master/002-es6-modules.md) that invokes a lot of thoughts relevant to this.

**Comment by rhettlivingston** [comment]
Hacked a bit on regex101.com and have a good start on a regex for extracting export information. You can play with it online at https://regex101.com/r/hR8sW1/4.
```
# The goal is to gather all ExportNames and ModuleRequests from the
# export statements in an es7 module. Another stage will be required
# to extract ExportNames from any recognized exportsList or
# bindingList.
#
# This is not written for validation. It is permissive. For
# example, it would accept 
#   "export *;" and 
#   "export const not_initialized;"
#
# Stripping comments beforehand would increase reliability.
#
# There are some areas where obscure stuff will make it miss exported
# variable names. For example,
#    "export const name1 = 'some string ;', name2 = 'another';
# where the first semicolon will cause it to stop and miss name2.
# Stripping string literals down to '' before running it could help
# that and many other fringe cases. That could be done in the same
# pass that strips comments.

\bexport\b\s*                        # <declarationKeyword>
(?:(?:var|let|const)\b\s*
     (?<bindingList>[^;]*)           # <bindingList>
|  (?<default>default)               # <default> is all we need
|  function\s+    
     (?<functionName>[a-zA-Z$_]\w*)  # <functionName>
|  function\s*\*\s*
     (?<generatorName>[a-zA-Z$_]\w*) # <generatorName>
|  class\s+
     (?<className>[a-zA-Z$_]\w*)     # <className>
| (?:
    (?<importName>\*)|               # <importName> = '*'
    {\s*(?<exportsList>[^}]*?)\s*}   # <exportsList>
  )\s*
  (?:
    from\s*
    (?<quote>'|\")                   # <quote> opening quote
    (?<moduleSpecifier>[^\n]+?)      # <moduleSpecifier>
    \k<quote>\s*                     # closing quote
  )?
)
```

**Comment by lencioni** [comment]
@rhettlivingston that looks like a pretty good start on a regex based approach. I noticed that it grabs things from inside comments and strings, which is probably not what we want. For example, it will match on these:

```
// export default foo;
"export default foo;"
```

We won't be able to avoid all of these types of cases with a simple regex solution (e.g. multi-line comments like `/* ... */` might be challenging without a real parser), but we can probably tweak your regex to avoid some simple cases like I mentioned here.

**Comment by lencioni** [comment]
Specifically, I think you can change the first part to `^\s*export\b\s*`, since I don't think anything other than whitespace can precede an `export`.

**Comment by rhettlivingston** [comment]
I'm really thinking that the only way to reliably handle the comments and the issues with what could be inside quotes in an initializer is to have a two-pass process that strips them beforehand. The recognizer for the first pass would be very simple and fast compared to the one above.

>Specifically, I think you can change the first part to ^\s*export\b\s*, since I don't think anything other than whitespace can precede an export.

Though unconventional, I didn't think JS cared about newlines that much. Isn't
`export this; export that; export andThat;` all on one line legal and in fact might even be encountered if some package had been minified or something?

As I've gotten into this, the most difficult part seems to be writing a regex to parse the "bindingList" returned from the above regex. It has [this syntax](https://tc39.github.io/ecma262/#prod-BindingList). The problem is figuring out how to skip the "initializer"s. You can't just look for the next comma because the initializers can include things like arrow functions, new expressions, etc. that often have commas. This may be the aspect that drives me to babylon. Exporting an arrow function, for example, is not that unusual, and, yes, I've seen cases of exporting lists of arrow functions.

**Comment by rhettlivingston** [comment]
Just a "little" update (partly to gather my own thoughts)...

I'm now working with babylon and an AST for a take 2 attempt. I've chosen use babylon directly instead of via babel because I only need an AST and there is a feature I can't turn on in babylon through babel.

I've built something that can extract imports/exports from an ES module (working on dropping the "6" from my vocabulary because there are already extensions to ES6 modules) with a few caveats.

First caveat. You can't parse just "an" ES module and fully determine its exports. This is spelled out most clearly in step 7 of [the specification for the GetExportedNames concrete method](https://tc39.github.io/ecma262/#sec-getexportednames) where GetExportedNames calls HostResolveImportedModule and then recurses to get the exported names of that module. Thus, to get the full set of exported names from a module, you may have to recurse to other modules, packages, core, etc. This is the result of the ES ExportAllDeclaration, `export * from 'x';`, which re-exports all names exported by 'x' from the current module. The only way you can find what is being exported is thus to go to the source.

This leads to the second caveat, HostResolveImportedModule. It's called "Host" because it is implementation specific, i.e. node's require and similar things in every environment you come across. Note that though this is related to "require.resolve" and similar things, it **returns a "Module Record"**. This is a very important distinction! I'm feeling that we need to let environments provide their own HostResolveImportedModule and to possibly make use of the default one. We also need a strong, clean default one. This caveat is the root of the thing that could upend everything. It hearkens back to the TL;DR post I made in #329 the other day. The "modules" configuration thing I mentioned at the end of that would directly provide objects that HostResolveImportedModule could return.

And finally, the third caveat or rather same-old-can-of-worms. That resolution process can and will lead to non-module things - most often something that is CommonJS. And CommonJS exports can't be reliably statically analyzed (though I may take a shot at it at some point since we don't require perfection). Which is the whole reason we have things like namedExports, aliases, coreModules, etc.

i.e. HostResolveImportedModule needs to be able to resolve to a Module object derived from our options in addition to source code and package descriptor (Meteor's package.js can define exports) sources.

i.e. this new thing I'm making wants to evolve to be everything we do in the discovery phase :disappointed: I didn't see this clearly when I started,,, but I think it could clean up a lot of confusion in the process.

Another thought thread comes from connecting the above with our application needs and thinking about performance. A compiler working with EcmaScript never has to compile anything that isn't needed. It starts at a point and the imports lead it to exactly what it needs. We can't do that. Our starting point is assumed to be missing imports. The variable names cannot lead us to the things we need because

1. even if they represent a default export they don't always correspond in any way to the filename of what we need, 
2. they can correspond to multiple files in different locations of the same name, and 
3. they can correspond to namedExports. So, we're going to have to scan everything in reach before making our decisions, not just what the code is leading us to.

To work right, we'll need to build a Modules pool in phases. Something like
 - create Modules to represent all core modules (not just their names, but what they export)
 - create Modules to represent user or environment defined modules (directly specified ones, not parser derived ones)
 - get the environments to create Modules to represent all packages contained in whatever form(s) of dependencies the environment has (here we go a-parsing)
 - create Modules for all local modules / scripts we can find (more parsing)

When that is done, we'll have all possible information to perform whatever has been asked of us: fix, word, goto, etc. The Modules facility will have tools to find all EsModule objects that export a particular variable, etc. Note that those finds will never be looking into files or directories, that is done by the process building the Modules pool.

A mile-high overview of my suspected destination is that I'm heading for what is "suggested" by the EcmaScript specification in 15.2 - an EsModules class that is a container for EsModule (abstract) objects that include EsModuleModule (concrete) objects and EsModulePredefined (concrete) objects. I'm also suspecting EsModuleMeteor (concrete) will join that. I don't know about EsModuleNpm since npm has no interface definition of its own. EsModuleCommonJS (concrete) is a possibility though it would always suffer some reliability issues that drive the user to feed more into EsModulePredefined. All of those concrete objects essentially just define what the "thing" they represent exports in a common way. Of the above-mentioned possibilities, I only have EsModuleModule at a 90% level but it keeps getting hit by refactoring.

Anyway, I'll check something in soon. It is highly speculative; will start with only EsModules, EsModule, and EsModuleModule; and may be test-driven only (not hooked into the rest of our codebase yet in any way) when I check it in.

**Comment by trotzig** [comment]
Thanks for sharing your findings. I'm excited to see what you come up with!

**Comment by trotzig** [comment]
This will be fixed as soon as #365 is merged.

**Comment by trotzig** [comment]
With the 2.1 release, we have the basic support needed to find named exports. Closing.
